File: jas_ocaml/lib/menu/menu_model.ml
(** Menu bar render model — projected from the compiled workspace
    [menubar] (menubar.yaml).

    The menu bar is rendered (in [Menubar]) from [menu_bar_model], which
    projects the single source of truth — the compiled [menubar] — into a
    render model. This replaced a hand-maintained set of five hardcoded
    native menus that had drifted from the spec (e.g. the View menu was
    missing Actual Size / Fit Artboard / Fit All, Save and Save As both
    bound Ctrl+S). Projecting from the bundle means the OCaml menu bar can
    no longer diverge from menubar.yaml.

    The dynamic Workspace / Appearance submenus stay runtime-populated by
    bespoke code in [Menubar]; the model only carries their trigger label
    and identity. This module is GTK-free so it can be unit-tested
    headless. Mirrors the Rust menu.rs projector. *)

(** Which runtime-populated submenu a [Dynamic_submenu] entry drives. *)
type submenu_kind = Workspace | Appearance

(** One resolved menu entry. *)
type entry =
  | Separator
  | Dynamic_submenu of { label : string; kind : submenu_kind }
  | Action of {
      label : string;
      action : string;
      params : (string * Yojson.Safe.t) list;
      shortcut : string;
      enabled_when : string option;
    }

(** One top-level menu (e.g. "&File") and its entries. *)
type menu = { label : string; entries : entry list }

let str_member key j =
  match Workspace_loader.json_member key j with
  | Some (`String s) -> s
  | _ -> ""

let contains_sub (s : string) (sub : string) : bool =
  let n = String.length s and m = String.length sub in
  if m = 0 then true
  else
    let rec loop i =
      if i + m > n then false
      else if String.sub s i m = sub then true
      else loop (i + 1)
    in
    loop 0

let project_entry (item : Yojson.Safe.t) : entry =
  match item with
  (* A bare "separator" string. *)
  | `String "separator" -> Separator
  | _ ->
    (* A submenu carries nested "items"; the only ones today are the
       dynamic Workspace / Appearance submenus, rendered natively. *)
    (match Workspace_loader.json_member "items" item with
     | Some _ ->
       let label = str_member "label" item in
       let id = str_member "id" item in
       let kind =
         if contains_sub id "appearance" || contains_sub label "Appearance"
         then Appearance
         else Workspace
       in
       Dynamic_submenu { label; kind }
     | None ->
       let params =
         match Workspace_loader.json_member "params" item with
         | Some (`Assoc pairs) -> pairs
         | _ -> []
       in
       let enabled_when =
         match Workspace_loader.json_member "enabled_when" item with
         | Some (`String s) -> Some s
         | _ -> None
       in
       Action {
         label = str_member "label" item;
         action = str_member "action" item;
         params;
         shortcut = str_member "shortcut" item;
         enabled_when;
       })

let project_menu (menu : Yojson.Safe.t) : menu =
  let label = str_member "label" menu in
  let entries =
    match Workspace_loader.json_member "items" menu with
    | Some (`List items) -> List.map project_entry items
    | _ -> []
  in
  { label; entries }

let menu_bar_model () : menu list =
  match Workspace_loader.load () with
  | None -> []
  | Some ws -> List.map project_menu (Workspace_loader.menubar ws)

let strip_mnemonic (label : string) : string =
  let buf = Buffer.create (String.length label) in
  let n = String.length label in
  let i = ref 0 in
  while !i < n do
    let c = label.[!i] in
    if c = '&' then begin
      if !i + 1 < n && label.[!i + 1] = '&' then begin
        (* Escaped literal ampersand. *)
        Buffer.add_char buf '&';
        i := !i + 2
      end else
        (* Drop the mnemonic marker. *)
        incr i
    end else begin
      Buffer.add_char buf c;
      incr i
    end
  done;
  Buffer.contents buf

File: jas/menu/menu_model.py
"""Menu bar data — projected from the compiled workspace ``menubar``.

The top menu bar is rendered (in ``menu.py``) from :func:`build_menu_model`,
which projects the single source of truth — the compiled ``menubar``
(menubar.yaml, bundled into workspace.json) — into a render model. This
replaced a hand-maintained native menu that had drifted from the spec (it
carried stale shortcuts, an extra Delete item, and was missing the Actual
Size / Fit Artboard / Fit All view entries and most of the Window panel
toggles). Projecting from the bundle means the Python menu bar can no longer
diverge from menubar.yaml.

The dynamic Workspace / Appearance submenus stay runtime-populated by bespoke
code in ``menu.py``; the model only carries their trigger label and identity
(:class:`MenuSubmenu`). Mirrors the Rust ``menu.rs::menu_bar_model`` projector.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum, auto


class SubmenuKind(Enum):
    """Which runtime-populated submenu a :class:`MenuSubmenu` drives."""

    WORKSPACE = auto()
    APPEARANCE = auto()


@dataclass
class MenuSeparator:
    """A menu separator (a bare ``"separator"`` string in the bundle)."""


@dataclass
class MenuSubmenu:
    """A dynamic submenu (Workspace / Appearance) — runtime-populated by
    bespoke native code; the model only carries its label + kind."""

    label: str
    kind: SubmenuKind


@dataclass
class MenuAction:
    """One resolved action entry.

    ``label`` keeps the ``&`` mnemonic markers verbatim (Qt consumes ``&`` as
    its native accelerator marker). ``enabled_when`` is carried for fidelity
    but NOT evaluated here — enable/disable stays native (same as Rust v1).
    """

    label: str
    action: str
    params: dict = field(default_factory=dict)
    shortcut: str = ""
    enabled_when: str | None = None


@dataclass
class MenuModel:
    """One top-level menu (e.g. ``"&File"``) and its entries."""

    label: str
    entries: list = field(default_factory=list)


def build_menu_model() -> list[MenuModel]:
    """Project the compiled ``menubar`` (menubar.yaml) into the render model.

    Returns an empty list if the bundle is missing/corrupt (never raises).
    """
    from panels.yaml_menu import get_workspace_data

    ws = get_workspace_data()
    if not ws:
        return []
    menus = ws.get("menubar")
    if not isinstance(menus, list):
        return []
    return [_project_menu(m) for m in menus if isinstance(m, dict)]


def _project_menu(menu: dict) -> MenuModel:
    label = menu.get("label", "") or ""
    entries = []
    for item in menu.get("items", []) or []:
        entries.append(_project_entry(item))
    return MenuModel(label=label, entries=entries)


def _project_entry(item):
    # A bare "separator" string.
    if item == "separator":
        return MenuSeparator()
    if not isinstance(item, dict):
        return MenuSeparator()
    # A submenu carries nested "items"; the only ones today are the dynamic
    # Workspace / Appearance submenus, rendered natively (runtime-populated).
    if "items" in item:
        label = item.get("label", "") or ""
        ident = item.get("id", "") or ""
        if "appearance" in ident or "Appearance" in label:
            kind = SubmenuKind.APPEARANCE
        else:
            kind = SubmenuKind.WORKSPACE
        return MenuSubmenu(label=label, kind=kind)
    params = item.get("params")
    if not isinstance(params, dict):
        params = {}
    return MenuAction(
        label=item.get("label", "") or "",
        action=item.get("action", "") or "",
        params=params,
        shortcut=item.get("shortcut", "") or "",
        enabled_when=item.get("enabled_when"),
    )
